### DISCUSSION TITLE
internationalization (i18n) / multiple language support

### DISCUSSION THREAD

**Discussion by eliandoran** [TriliumNext/Notes#240] [Development]
One of the core aspects of Trilium that is missing is the fact that it is an English-only software.

There are solutions such as @Nriver 's repo which work by replacing the hard-coded messages within the application via regular expressions.

Speaking of which, @Nriver , would you be willing to help us with setting up Chinese for the application?

On my side I would prefer a different method which involves adding multilanguage support directly in the application instead of doing it at build time, however your existing expertise plus translations would go a long way.

What do you think?

**Comment by zerebos** [comment]
> On my side I would prefer a different method which involves adding multilanguage support directly in the application instead of doing it at build time

I absolutely agree with this, it would be good for both us as well as the users in the end. We can make use of something like `i18-next` to help handle edge cases and a crowd translation tool to generate locale files as well as solicit translations from anyone willing to contribute.

**Comment by Nriver** [comment]
Ok, let me take a look at this. It will take some time.

I know you have just finished transitioning from JS to TS. Thank you for your hard work over the past months. Kudos to you for that :)

The search feature is blazing fast in the new version, which is a significant improvement compared to the original one. I really like it and feel more motivated to switch to the new version now.

I have a question about contributing to the code here. I previously forked the original repo from Zadam, and I tried to fork this repo now, but I encountered this error message:
![image](https://github.com/user-attachments/assets/18f71270-4660-49bb-bdb4-9f0dfd041e21)
What should I do? Should I delete my previous fork and create a new one from this repo, or is there a magical git command that can resolve this?

**Reply by zerebos** [reply]
What I would do is add the TriliumNext one as a second remote in your local repository, then you can pull changes from here and submit them here as well.

**Reply by Nriver** [reply]
Thanks for the advice about the secondary upstream.
I can pull the changes from here, but I do not have permission to make a direct push : (
But I made a push to my fork and made a PR on Github and specify to push to TriliumNext here : ) #245

**Reply by eliandoran** [reply]
@Nriver , we can keep working like this for now. After a few PRs I will give you full dev access.
Anyway, there's not much of a difference regarding using a fork vs pushing branches directly.

**Comment by Nriver** [comment]
At present, `i18next` appears to be the optimal choice for handling translations in a Node.js project. It seems capable of managing most of the UI text for Trilium effectively.

However, based on my previous experience translating Trilium into Chinese, I want to highlight a few important points:

1. **CKEditor Language File**
   - Trilium's built-in CKEditor is hardcoded and does not include a language file. I had to replace it with a translated version. See [here](https://github.com/Nriver/trilium-translation/blob/5803b6ecab05bf42532988e515751409923725c2/2.trans.py#L3326-L3346). So we need to include more language and make it configuable somehow.

2. **Demo Document Translation**
   - The demo document `db/demo.zip` is currently in English. It would be reasonable to create multiple language-specific demos to support i18n.

3. **Excalidraw Language Configuration**
   - Excalidraw has its own built-in language file, which needs to be made configurable. The `langCode` parameter is available in the initialization process. See the relevant code [here](https://github.com/Nriver/trilium-translation/blob/5803b6ecab05bf42532988e515751409923725c2/2.trans.py#L3265-L3281).

4. **Excalidraw Font Replacement**
   - Excalidraw's handwriting font is hardcoded. I had to replace it with a font that supports Chinese character. This could pose a problem if additional language support is required. Relevant code can be found [here](https://github.com/Nriver/trilium-translation/blob/5803b6ecab05bf42532988e515751409923725c2/2.trans.py#L3351-L3358) and [here](https://github.com/Nriver/trilium-translation/blob/5803b6ecab05bf42532988e515751409923725c2/2.trans.py#L2633-L2866).

That's it for now. Starting next week, I plan to use `i18next` for translating Trilium.

**Reply by eliandoran** [reply]
@Nriver , this is a wonderful analysis, thank you.

Regarding `i18next`, would you need help integrating it? Regarding CKEditor we were planning to extract it out of the repository so we might as well do something about the translation packs as well.

Regarding Excalidraw I"ll check if there is any upstream support for changing the font.

**Reply by Nriver** [reply]
Yes, it would be great if you could start with a simple demo of `i18next` integration for me. I'm not a professional in Node.js or TypeScript and I'm concerned that I might disrupt the code style.

**Reply by eliandoran** [reply]
@Nriver , sounds good. I'll look into it. I hope to have a PoC by the end of the week.

**Reply by eliandoran** [reply]
@Nriver , the first PoC for i18n is available in #248 .

I've attached some documentation in there.
Generally we can now translate any client-side message, apart from the ones in views. Server-side translations are not yet implemented.

The first step is to look at all the .js files in the client and replace the hard-coded messages with the translation mechanism as described in the documentation file.

Since you already did most of the work of translating and identifying the strings to translate, I would suggest making some kind of script that takes as an input the translation repository on your side and spits out a JSON file, plus changes the translations to the `t()` mechanism.

**Reply by Nriver** [reply]
I'll take a look and see what I can do.

**Reply by Nriver** [reply]
I'll continue working on #257

**Comment by hasecilu** [comment]
Hi, I'm interested to help on the translation to Spanish, I know the existence of the project with support for Chinese language and I was wondering if it's recommendable to translate in the same way or it's better to wait for i18n support arrive to the main project?

Also, thinking on translators, translating directly in source file is kind of clunky but doable, if it's possible to extract the strings to an external `.po` file, for example, would be much easier for translators to use a proper software (like [Poedit](https://poedit.net/)) to help on this task.

In order to get a better participation from people I think it would be wise to use an online translation tool where translators can directly translate with very nice UI. My personal favorite (as translator) is [Weblate](https://weblate.org) but there are other tools available. Weblate is free for FOSS. What do you think?

**Reply by eliandoran** [reply]
@hasecilu , yes, we'll first set up a proper system for translations first and maybe even a translation platform like Weblate or Crowdin. :)

**Reply by eliandoran** [reply]
I will let you know once we manage to make progress.

**Reply by kozmotronik** [reply]
> @hasecilu , yes, we'll first set up a proper system for translations first and maybe even a translation platform like Weblate or Crowdin. :)

I've contributed on some projects in Crowdin and would be eager to contribution for this project, too.

> In order to get a better participation from people I think it would be wise to use an online translation tool where translators can directly translate with very nice UI.

I totally agree with that. The app would receive more translation contributions comparing to forking the whole repo to contribute for the translation. Weblate is FOSS and also self-hostable.

So how is this process going on? Have you managed to switch to any internationalization tool?

**Reply by Potjoe-97** [reply]
I manage French Translation with Crowdin, it is a great tool.

**Comment by meichthys** [comment]
See #338
